### PR TITLE
feat: Replication testing

### PR DESCRIPTION
**Pull request by richburdon**
### Summary
🛠️🚀🧹

This pull request adds debugging features and UI improvements to the `labs-app`, `plugin-client`, and `plugin-debug` packages, using the `config`, `log`, and `protocols` modules from the SDK. It also updates the `feed-store` and `echo-pipeline` packages to fix some issues and enhance the performance and reliability of the feed and data management. Additionally, it updates some TypeScript configurations and dependencies, and removes some unnecessary whitespace and comments.

> _`plugin-client` grows_
> _adding `client` and `log`_
> _debug identity_

### Walkthrough
*  Enable debugging features for identity management and invitations in the `labs-app` and `plugin-client` packages ([link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-9d663519e7a46a19fe4305c8085aa017f51355119fb0fbe545eb37ab8492722cL34-R38),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-9d663519e7a46a19fe4305c8085aa017f51355119fb0fbe545eb37ab8492722cL48-R54),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-9d663519e7a46a19fe4305c8085aa017f51355119fb0fbe545eb37ab8492722cL60-R63),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-4789b3ee5b3e63f0181c552c1e9723e927d3e9a1cdca65aa1109c542fc087d17L18-R21),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-2980bfaacf44efa379f06240cc85cd206bc5b09d919e1ecfb8b170c6c4d08cfdL7-R10),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-2980bfaacf44efa379f06240cc85cd206bc5b09d919e1ecfb8b170c6c4d08cfdL31-R36),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-2980bfaacf44efa379f06240cc85cd206bc5b09d919e1ecfb8b170c6c4d08cfdR48-R67),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-e56910f2262fdeae3091d92ab8065942ae5b436d28156b7b554e0a37c9c88132L24-R32))
*  Add a button to create and copy invitations for spaces in the `plugin-debug` package, and use the `InvitationEncoder` from the SDK to encode and decode invitations ([link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-9d7c761af6f52ac3088160c56da2d94a888f41bd030fd3baaf75b50abcb06049R26),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-3af2dcc1d2d7818d416feb3e49aceb9c1abd5ca1857f5cbdc4054708e5ef1372L5-R5),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-3af2dcc1d2d7818d416feb3e49aceb9c1abd5ca1857f5cbdc4054708e5ef1372L17-R20),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-3af2dcc1d2d7818d416feb3e49aceb9c1abd5ca1857f5cbdc4054708e5ef1372L23-R27),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-3af2dcc1d2d7818d416feb3e49aceb9c1abd5ca1857f5cbdc4054708e5ef1372L56-R90),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-3af2dcc1d2d7818d416feb3e49aceb9c1abd5ca1857f5cbdc4054708e5ef1372L89-R116),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-ab69dba064213c83240eccb559dcfd6442d23cbd457d05b56880dcdc61c803bbL15-R15),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-6ffbb2ecffc00e13af28b2b955cc75a7cf3efc6d567e5b61f04a8dd8290fe9b7L27-R27),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-6ffbb2ecffc00e13af28b2b955cc75a7cf3efc6d567e5b61f04a8dd8290fe9b7L33-R37),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-e82ac13a9540bc9a4a66f84727936d46e6d0e183936baeecea6c29fa94148dd7R42-R44))
*  Add input fields to specify the number of mutations and the time interval for the debugging panel in the `plugin-debug` package, and limit the number of mutations accordingly ([link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-9c210751898e82a6bd5c22e15a24db5f3ffe6eebeb547cddd8ce90016a38a490L6-R6),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-9c210751898e82a6bd5c22e15a24db5f3ffe6eebeb547cddd8ce90016a38a490L29-R49),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-3af2dcc1d2d7818d416feb3e49aceb9c1abd5ca1857f5cbdc4054708e5ef1372R34-R36),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-3af2dcc1d2d7818d416feb3e49aceb9c1abd5ca1857f5cbdc4054708e5ef1372R48),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-3af2dcc1d2d7818d416feb3e49aceb9c1abd5ca1857f5cbdc4054708e5ef1372L98-R137),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-3af2dcc1d2d7818d416feb3e49aceb9c1abd5ca1857f5cbdc4054708e5ef1372L107-R152))
*  Improve the UI of the debugging panel in the `plugin-debug` package by using icons instead of text for the buttons ([link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-3af2dcc1d2d7818d416feb3e49aceb9c1abd5ca1857f5cbdc4054708e5ef1372L5-R5),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-3af2dcc1d2d7818d416feb3e49aceb9c1abd5ca1857f5cbdc4054708e5ef1372L89-R116))
*  Increase the batch size for reading feeds in batches in the `feed-store` package, and use the `defaultsDeep` function to merge hypercore options ([link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-b61df366c52f452d49c6a479e47aef3c33a642e65c87da080208a5d330f79320R35),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-b61df366c52f452d49c6a479e47aef3c33a642e65c87da080208a5d330f79320L40-R42),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-bfb339b43f46301885ddb880d1ae51608ec169343465aef336751501a4953758R5),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-bfb339b43f46301885ddb880d1ae51608ec169343465aef336751501a4953758L71-R75),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-a8b9cf688aad506de3104c55cc8ff78bea1ac77de77007fe8d08f68e003e2facL19-R19))
*  Make the `close` method of the `Stream` class in the `codec-protobuf` package asynchronous, and await the `dispose` method ([link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-aa33e2bcebcdb60fd181e7c772325a681b87e96ba7bf95b2005b6f43f94bb9beL277-R277),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-aa33e2bcebcdb60fd181e7c772325a681b87e96ba7bf95b2005b6f43f94bb9beL286-R285))
*  Log the hypercore options and the feed key when opening feeds in the `feed-store` package, and disable the sparse option for now ([link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-bfb339b43f46301885ddb880d1ae51608ec169343465aef336751501a4953758L42-R46),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-bfb339b43f46301885ddb880d1ae51608ec169343465aef336751501a4953758R56-R57),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-0344ff6fdadd44711572efd369edcc0a4d132baa4ff154b6bbb195c505d6d00eR58-R62),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-0344ff6fdadd44711572efd369edcc0a4d132baa4ff154b6bbb195c505d6d00eL78-R87),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-64fa5ed50276a5b9c275e0a0c5783bf20564f732925405ced87a10f9fd6dc8d5L218-R227))
*  Log the feed key and length when the feed is downloaded in the `echo-pipeline` package, and use the `log.error` method for the error case ([link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-1df1310cca796382869c882d8f953732c5ea557af4f9e4219223c7eada3fb9b3L375-R383))
*  Clarify the comment for the `onwrite` hook in the `feed-store` package, and remove redundant or unnecessary code and whitespace ([link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-bfb339b43f46301885ddb880d1ae51608ec169343465aef336751501a4953758L26-R27),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-32a8f10b8b69f902a8079509c132751bd3f4cac5c1718a031927188faf627aaeR125))
*  Add a TODO comment and fix a formatting error in the `todomvc` package ([link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-f1a95f934820faae0bfd5466994df87394441875ff6ad597e6d969936d14431fR12),[link](https://github.com/dxos/dxos/pull/3972/files?diff=unified&w=0#diff-f1a95f934820faae0bfd5466994df87394441875ff6ad597e6d969936d14431fL29-R30))

### CODE REVIEW

File: packages/common/feed-store/src/feed-store.ts [hunk 1]
```diff
@@ -55,14 +55,23 @@ export class FeedStore<T extends {}> {
     invariant(feedKey);
     invariant(!this._closed, 'Feed store is closed');
 
+    // TODO(burdon): Coordinate with FeedFactory!!!
+    if (sparse) {
```
Comment (dmaretskyi) [inline packages/common/feed-store/src/feed-store.ts:59]: can you walk me throught the rationale here?

File: packages/core/mesh/network-manager/src/transport/webrtc-transport.ts [hunk 2]
```diff
@@ -17,31 +17,46 @@ import { wrtc } from './webrtc';
...
-
   constructor(private readonly params: WebRTCTransportParams) {
     log.trace('dxos.mesh.webrtc-transport.constructor', trace.begin({ id: this._instanceId }));
     log('created connection', params);
+
+    // https://www.npmjs.com/package/simple-peer#api
+    // NOTE: Set localStorage.debug = 'simple-peer*' to enable debug logging. (Or show verbose in Chrome).
+    // View throughput via: chrome://webrtc-internals (data-channel)
     this._peer = new SimplePeerConstructor({
+      channelName: this._instanceId,
```
Comment (dmaretskyi) [inline packages/core/mesh/network-manager/src/transport/webrtc-transport.ts:51]: wouldn't this break the networking? I though channel names needed to match between peers

File: packages/core/mesh/network-manager/src/swarm/peer.ts [hunk 3]
```diff
@@ -81,6 +81,8 @@ export class Peer {
   public advertizing = false;
 
   public initiating = false;
+
+  // TODO(burdon): Convert to map.
```
Comment (dmaretskyi) [inline packages/core/mesh/network-manager/src/swarm/peer.ts:85]: you mean the constructor options?

File: packages/core/echo/echo-pipeline/src/pipeline/pipeline.ts [hunk 4]
```diff
@@ -372,11 +372,15 @@ export class Pipeline implements PipelineAccessor {
...
   private _setFeedDownloadState(feed: FeedWrapper<FeedMessage>) {
     const timeframe = this._state._startTimeframe;
-    const seq = timeframe.get(feed.key) ?? 0;
-
-    feed.undownload({ callback: () => log('undownload') });
-    feed.download({ start: seq + 1, linear: true }).catch((err: Error) => {
-      log('failed to download feed', { err });
+    const seq = timeframe.get(feed.key) ?? -1;
+
+    feed.undownload({
```
Comment (dmaretskyi) [inline packages/core/echo/echo-pipeline/src/pipeline/pipeline.ts:377]: hmm, is this better then calling them in parallel? I'm concerned about possible race conditions if _setFeedDownloadState is called again before undownload callback could complete

**Review by dmaretskyi — CHANGES_REQUESTED** [review]
in comments

File: packages/core/echo/echo-pipeline/src/pipeline/pipeline.ts (hunk 4 above)
Comment (richburdon) [reply packages/core/echo/echo-pipeline/src/pipeline/pipeline.ts:399]: ok good point

File: packages/core/mesh/network-manager/src/swarm/peer.ts (hunk 3 above)
Comment (richburdon) [reply packages/core/mesh/network-manager/src/swarm/peer.ts:85]: yes

File: packages/core/mesh/network-manager/src/transport/webrtc-transport.ts (hunk 2 above)
Comment (richburdon) [reply packages/core/mesh/network-manager/src/transport/webrtc-transport.ts:51]: i thought it was decorative. i'll check.

**Pull request merged**